crates/valence_nbt/src/list.rs

```diff
@@ -388,6 +388,44 @@ impl List {
             List::Compound(list) => Value::Compound(list.remove(index)),
             List::IntArray(list) => Value::IntArray(list.remove(index)),
             List::LongArray(list) => Value::LongArray(list.remove(index)),
+        };
+
+        if self.is_empty() {
+            *self = List::End;
+        }
+
+        removed
+    }
+
+    /// Returns only the elements specified by the predicate, passing a mutable
+    /// reference to it.
+    ///
+    /// In other words, removes all elements `e` such that `f(ValueMut(&mut e))`
+    /// returns `false`. This method operates in place, visiting each element
+    /// exactly once in the original order, and preserves the order of the
+    /// retained elements.
+    pub fn retain<F>(&mut self, mut f: F)
+    where
+        F: FnMut(ValueMut) -> bool,
```

Comment:
We shouldn't borrow the values in the list as mutable.

Reply:
Why not?

Reply:
It generally bad practice to make things mutable when they don't need to be, and I feel like we'd be encouraging some messy antipatterns. The way `retain` works is it uses a predicate which should make assertions about the value it's given in a read-only manor.

The [standard library does this](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain). It delegates the mutable version to `retain_mut`.

Reply:
See the explanation in the OP

Reply:
I'll have to defer this to @rj00a

Reply:
I agree with @Earthcomputer on this one.